### PR TITLE
Pyramid level customization

### PR DESCRIPTION
**Pull request by mooratov**
TLDR, This feature accomplishes the following:
- allows user to pass custom list of pyramid levels via config file
- these pyramid levels are then parsed in all appropriate places. 
- allows use of pyramid layer 2 (P2) for resnet backbone 
- one minor unrelated bug fix in `debug.py` ( made`--show-annotations` a separate flag from `--annotations`, which is already in use) 

This PR would allow a user to use a custom set of pyramid levels, controlled by the config.ini file in a manner similar to how anchor_parameters are controlled now. 

Our team has found that using a P2 pyramid layer enables us to utilize ultra-fine (e.g. 3 pixel by 3 pixel) bounding boxes that are otherwise out of reach with current configurable parameters. Some of the code used here was contributed by my team member Brad Nelson (https://github.com/bnelsj) in our private fork.  

I tried to stick to stylistic conventions wherever I could, and it looked like `utils/anchors.py` already had a framework for configuring the `pyramid_levels` list. I simply implemented this more fully in the bin scripts. Unfortunately, there was a fair amount of re-writing the `retinanet.py` definition and making it somewhat uglier.  

In order to make P2 work, I had to modify `resnet.py` so that it would return 4 output layers to use as backbone, while no other backbone type currently returns 4 output layers. Therefore, pyramid level 2 (P2) only works with resnet backbones as written. I tried to explain these changes with informative errors and comments. 

The primary use of this feature would be to add P2, although a user can also exclude P7 and P6 if they desired. I did not allow users to exclude C3, C4, and C5. 

Note that the length of `pyramid_levels` must be the same as the list of `sizes` and `strides` in anchor_parameters, although this is not explicitly enforced.

### CODE REVIEW

File: keras_retinanet/models/retinanet.py [hunk 1]
```diff
@@ -124,42 +124,68 @@ def default_regression_model(num_values, num_anchors, pyramid_feature_size=256,
...
+        P3_upsampled = layers.UpsampleLike(name='P3_upsampled')([P3, C2])
     P3 = keras.layers.Conv2D(feature_size, kernel_size=3, strides=1, padding='same', name='P3')(P3)
+    output_layers["P3"] = P3
+
+    if C2 is not None and 2 in pyramid_levels:
+        P2 = keras.layers.Conv2D(feature_size, kernel_size=1, strides=1, padding='same', name='C2_reduced')(C2)
+        P2 = keras.layers.Add(name='P2_merged')([P3_upsampled, P2])
+        P2 = keras.layers.Conv2D(feature_size, kernel_size=3, strides=1, padding='same', name='P2')(P2)
+        output_layers["P2"] = P2
+
```
Comment (hgaiser) [inline keras_retinanet/models/retinanet.py:170]: Why these two newlines?

File: keras_retinanet/models/retinanet.py [hunk 2]
```diff
@@ -124,42 +124,68 @@ def default_regression_model(num_values, num_anchors, pyramid_feature_size=256,
...
+
 
     # "P6 is obtained via a 3x3 stride-2 conv on C5"
-    P6 = keras.layers.Conv2D(feature_size, kernel_size=3, strides=2, padding='same', name='P6')(C5)
+    if 6 in pyramid_levels:
+        P6 = keras.layers.Conv2D(feature_size, kernel_size=3, strides=2, padding='same', name='P6')(C5)
+        output_layers["P6"] = P6
+
```
Comment (hgaiser) [inline keras_retinanet/models/retinanet.py:177]: And here.

File: keras_retinanet/models/retinanet.py [hunk 3]
```diff
@@ -124,42 +124,68 @@ def default_regression_model(num_values, num_anchors, pyramid_feature_size=256,
...
-    P7 = keras.layers.Activation('relu', name='C6_relu')(P6)
-    P7 = keras.layers.Conv2D(feature_size, kernel_size=3, strides=2, padding='same', name='P7')(P7)
+    if 7 in pyramid_levels:
+        if 6 not in pyramid_levels:
+            raise ValueError("P6 is required to use P7")
+        P7 = keras.layers.Activation('relu', name='C6_relu')(P6)
+        P7 = keras.layers.Conv2D(feature_size, kernel_size=3, strides=2, padding='same', name='P7')(P7)
+        output_layers["P7"] = P7
+
+    return [output_layers['P'+str(p)] for p in pyramid_levels]
```
Comment (hgaiser) [inline keras_retinanet/models/retinanet.py:187]: Prefer to use `'P{}'.format(p)` for p in pyramid_levels`

File: keras_retinanet/models/retinanet.py [hunk 4]
```diff
@@ -273,10 +301,54 @@ def retinanet(
     if submodels is None:
         submodels = default_submodels(num_classes, num_anchors)
 
-    C3, C4, C5 = backbone_layers
+    if pyramid_levels is None:
+        pyramid_levels = [3,4,5,6,7]
```
Comment (hgaiser) [inline keras_retinanet/models/retinanet.py:305]: Spaces after the comma.

File: keras_retinanet/utils/config.py [hunk 5]
```diff
@@ -45,3 +45,8 @@ def parse_anchor_parameters(config):
     strides = list(map(int, config['anchor_parameters']['strides'].split(' ')))
 
     return AnchorParameters(sizes, strides, ratios, scales)
+
+def parse_pyramid_levels(config):
+    levels   = list(map(int, config['pyramid_levels']['P'].split(' ')))
```
Comment (hgaiser) [inline keras_retinanet/utils/config.py:50]: Too many spaces before the `=` sign.

File: keras_retinanet/preprocessing/generator.py [hunk 6]
```diff
@@ -313,9 +313,13 @@ def compute_inputs(self, image_group):
...
     def generate_anchors(self, image_shape):
         anchor_params = None
+        pyramid_levels = None
         if self.config and 'anchor_parameters' in self.config:
             anchor_params = parse_anchor_parameters(self.config)
-        return anchors_for_shape(image_shape, anchor_params=anchor_params, shapes_callback=self.compute_shapes)
+        if self.config and 'pyramid_levels' in self.config:
+            pyramid_levels = parse_pyramid_levels(self.config)
+
+        return anchors_for_shape(image_shape, anchor_params=anchor_params, pyramid_levels = pyramid_levels, shapes_callback=self.compute_shapes)
```
Comment (hgaiser) [inline keras_retinanet/preprocessing/generator.py:322]: No spaces around the `=` sign.

File: keras_retinanet/models/retinanet.py [hunk 7]
```diff
@@ -273,10 +301,54 @@ def retinanet(
...
+
+    backbone_layer_counter = 0
+    backbone_layer_number = len(backbone_layers)
+
 
+    if backbone_layer_number < 3:
+        raise ValueError('Must provide at least 3 layers for backbone')
+
+    if backbone_layer_number < 4 and 2 in pyramid_levels:
+        raise ValueError("backbone model providing {} layers. Need 4 backbone layers to use pyramid level 2 layers.\
```
Comment (hgaiser) [inline keras_retinanet/models/retinanet.py:318]: This kind of string concatenation doesn't work:
```python
In [1]: print("Hello \ 
   ...:     world")                                                                                                                                                                                                                            
Hello     world
```

File: keras_retinanet/models/retinanet.py [hunk 8]
```diff
@@ -273,10 +301,54 @@ def retinanet(
...
+         Only resnet returns 4 layers and will work with pyramid level 2 ".format(backbone_layer_number))
```
Comment (hgaiser) [inline keras_retinanet/models/retinanet.py:319]: I would leave this detail out, otherwise we need to keep updating this message when other models support it too.

File: keras_retinanet/models/retinanet.py [hunk 9]
```diff
@@ -273,10 +301,54 @@ def retinanet(
...
+
+    # Resent backbone has 4 usable layers. Other backbone models have 3.
```
Comment (hgaiser) [inline keras_retinanet/models/retinanet.py:321]: Same here, I'd leave this detail out.

Comment (mooratov) [reply keras_retinanet/models/retinanet.py]: alright, i took it out

File: keras_retinanet/bin/debug.py [hunk 10]
```diff
@@ -215,15 +215,15 @@ def run(generator, args, anchor_params):
                 image, image_scale = generator.resize_image(image)
                 annotations['bboxes'] *= image_scale
 
-            anchors = anchors_for_shape(image.shape, anchor_params=anchor_params)
+            anchors = anchors_for_shape(image.shape, anchor_params=anchor_params, pyramid_levels = pyramid_levels)
```
[the remark this answers is not in the document]
Comment (mooratov) [reply keras_retinanet/bin/debug.py]: alright fixed it

File: keras_retinanet/bin/train.py [hunk 11]
```diff
@@ -99,23 +99,27 @@ def create_models(backbone_retinanet, num_classes, weights, multi_gpu=0,
...
+        pyramid_levels = parse_pyramid_levels(config)
+        print('pyramid levels are', pyramid_levels)
 
     # Keras recommends initialising a multi-gpu model on the CPU to ease weight sharing, and to prevent OOM errors.
     # optionally wrap in a parallel model
     if multi_gpu > 1:
         from keras.utils import multi_gpu_model
         with tf.device('/cpu:0'):
-            model = model_with_weights(backbone_retinanet(num_classes, num_anchors=num_anchors, modifier=modifier), weights=weights, skip_mismatch=True)
+            model = model_with_weights(backbone_retinanet(num_classes, num_anchors=num_anchors, modifier=modifier,pyramid_levels=pyramid_levels), weights=weights, skip_mismatch=True)
```
[the remark this answers is not in the document]
Comment (mooratov) [reply keras_retinanet/bin/train.py]: fixed

File: keras_retinanet/bin/train.py [hunk 12]
```diff
@@ -99,23 +99,27 @@ def create_models(backbone_retinanet, num_classes, weights, multi_gpu=0,
...
         training_model = multi_gpu_model(model, gpus=multi_gpu)
     else:
-        model          = model_with_weights(backbone_retinanet(num_classes, num_anchors=num_anchors, modifier=modifier), weights=weights, skip_mismatch=True)
+        model          = model_with_weights(backbone_retinanet(num_classes, num_anchors=num_anchors, modifier=modifier, pyramid_levels=pyramid_levels), weights=weights, skip_mismatch=True)
         training_model = model
 
     # make prediction model
-    prediction_model = retinanet_bbox(model=model, anchor_params=anchor_params)
+    prediction_model = retinanet_bbox(model=model, anchor_params=anchor_params, pyramid_levels = pyramid_levels)
```
[the remark this answers is not in the document]
Comment (mooratov) [reply keras_retinanet/bin/train.py]: fixed

File: keras_retinanet/bin/train.py [hunk 13]
```diff
@@ -99,23 +99,27 @@ def create_models(backbone_retinanet, num_classes, weights, multi_gpu=0,
     # load anchor parameters, or pass None (so that defaults will be used)
     anchor_params = None
     num_anchors   = None
+    pyramid_levels = None
     if config and 'anchor_parameters' in config:
         anchor_params = parse_anchor_parameters(config)
         num_anchors   = anchor_params.num_anchors()
+    if config and 'pyramid_levels' in config:
+        pyramid_levels = parse_pyramid_levels(config)
+        print('pyramid levels are', pyramid_levels)
```
[the remark this answers is not in the document]
Comment (mooratov) [reply keras_retinanet/bin/train.py]: yeah sorry, removed now

File: keras_retinanet/bin/train.py [hunk 14]
```diff
@@ -479,9 +483,14 @@ def main(args=None):
         model            = models.load_model(args.snapshot, backbone_name=args.backbone)
         training_model   = model
         anchor_params    = None
+        pyramid_levels   = None
         if args.config and 'anchor_parameters' in args.config:
             anchor_params = parse_anchor_parameters(args.config)
-        prediction_model = retinanet_bbox(model=model, anchor_params=anchor_params)
+        if args.config and 'pyramid_levels' in args.config:
+            pyramid_levels = parse_pyramid_levels(args.config)
+            print('pyramid levels are', pyramid_levels)
```
[the remark this answers is not in the document]
Comment (mooratov) [reply keras_retinanet/bin/train.py]: yep, also removed now

File: keras_retinanet/models/resnet.py [hunk 15]
```diff
@@ -109,7 +109,8 @@ def resnet_retinanet(num_classes, backbone='resnet50', inputs=None, modifier=Non
         resnet = modifier(resnet)
 
     # create the full model
-    return retinanet.retinanet(inputs=inputs, num_classes=num_classes, backbone_layers=resnet.outputs[1:], **kwargs)
+    # resnet.outputs contains 4 layers
+    return retinanet.retinanet(inputs=inputs, num_classes=num_classes, backbone_layers=resnet.outputs[:], **kwargs)
```
[the remark this answers is not in the document]
Comment (mooratov) [reply keras_retinanet/models/resnet.py]: alright done.

File: keras_retinanet/models/retinanet.py [hunk 16]
```diff
@@ -124,42 +124,68 @@ def default_regression_model(num_values, num_anchors, pyramid_feature_size=256,
...
-def __create_pyramid_features(C3, C4, C5, feature_size=256):
+def __create_pyramid_features(C3, C4, C5, pyramid_levels, C2=None, feature_size=256):
     """ Creates the FPN layers on top of the backbone features.
 
     Args
         C3           : Feature stage C3 from the backbone.
         C4           : Feature stage C4 from the backbone.
         C5           : Feature stage C5 from the backbone.
+        C2           : Feature stage C2 from the backbone (if pyramid level 2 is in use).
+        pyramid_levels: pyramid levels in use.
```
[the remark this answers is not in the document]
Comment (mooratov) [reply keras_retinanet/models/retinanet.py]: done.

File: keras_retinanet/models/retinanet.py (hunk 1 above)
Comment (mooratov) [reply keras_retinanet/models/retinanet.py:170]: bad habit :). Removed now

File: keras_retinanet/models/retinanet.py (hunk 2 above)
Comment (mooratov) [reply keras_retinanet/models/retinanet.py:175]: removed also

File: keras_retinanet/models/retinanet.py (hunk 3 above)
Comment (mooratov) [reply keras_retinanet/models/retinanet.py]: fixed

File: keras_retinanet/models/retinanet.py (hunk 8 above)
Comment (mooratov) [reply keras_retinanet/models/retinanet.py]: alright done.

File: keras_retinanet/models/retinanet.py (hunk 7 above)
Comment (mooratov) [reply keras_retinanet/models/retinanet.py]: oh oops yeah. It's all one line now and since the non-resnet warnings are gone it's not too too long anymore

File: keras_retinanet/models/retinanet.py (hunk 4 above)
Comment (mooratov) [reply keras_retinanet/models/retinanet.py]: fixed

File: keras_retinanet/utils/config.py (hunk 5 above)
Comment (mooratov) [reply keras_retinanet/utils/config.py]: fixed

File: keras_retinanet/preprocessing/generator.py (hunk 6 above)
Comment (mooratov) [reply keras_retinanet/preprocessing/generator.py]: fixed

**Comment by mooratov** [comment]
Thank you so much for the detailed feedback! 
I believe I've made all the stylistic changes you requested. 

To answer your question, yes it would be fantastic if all backbone models returned dicts with "C" keys and layers as values, and FPN function would return dict with "P" keys! I was thinking of doing something like this, but figured it would be an even bigger overhaul and would require me to touch ALL of the models, many of which I haven't even tested out before. 
 
Are you asking me to make these changes as well prior to merge?

**Comment by hgaiser** [comment]
> Are you asking me to make these changes as well prior to merge?

Yeah if you don't mind, I think it would be a good change.

**Comment by mooratov** [comment]
OK I updated all backbone models, retinanet FPN function, and retinanet initialization all working with dicts.
I noticed that it was relatively easy to get densenet to work with P2 layers as well, so i enabled this capability. 

The other backbones only return C3, C4, and C5, as previously designed. Note that of the remaining models, I was only able to successfully run vggnet (keras 2.3.1, tf 1.14) but i don't believe this is due to any changes that I've made.

File: keras_retinanet/bin/convert_model.py [hunk 17]
```diff
@@ -66,11 +66,16 @@ def main(args=None):
...
     anchor_parameters = None
+    pyramid_levels = None
     if args.config:
         args.config = read_config_file(args.config)
         if 'anchor_parameters' in args.config:
             anchor_parameters = parse_anchor_parameters(args.config)
 
+        if 'pyramid_levels' in args.config:
+            pyramid_levels = parse_pyramid_levels(args.config)
+            print('pyramid levels are', pyramid_levels)
```
Comment (hgaiser) [inline keras_retinanet/bin/convert_model.py:77]: Debug print?

File: keras_retinanet/bin/evaluate.py [hunk 18]
```diff
@@ -154,7 +158,7 @@ def main(args=None):
 
     # optionally convert the model
     if args.convert_model:
-        model = models.convert_model(model, anchor_params=anchor_params)
+        model = models.convert_model(model, anchor_params=anchor_params, pyramid_levels = pyramid_levels)
```
Comment (hgaiser) [inline keras_retinanet/bin/evaluate.py:161]: No spaces.

File: keras_retinanet/models/densenet.py [hunk 19]
```diff
@@ -93,13 +93,21 @@ def densenet_retinanet(num_classes, backbone='densenet121', inputs=None, modifie
...
         model = modifier(model)
 
     # create the full model
-    model = retinanet.retinanet(inputs=inputs, num_classes=num_classes, backbone_layers=model.outputs, **kwargs)
+    backbone_layers = {
+        'C2' : model.outputs[0],
+        'C3' : model.outputs[1],
+        'C4' : model.outputs[2],
+        'C5' : model.outputs[3]
+        }
```
Comment (hgaiser) [inline keras_retinanet/models/densenet.py:109]: This line should be unindented once.

File: keras_retinanet/models/effnet.py [hunk 20]
```diff
@@ -117,8 +117,15 @@ def effnet_retinanet(num_classes, backbone='EfficientNetB0', inputs=None, modifi
     if modifier:
         model = modifier(model)
 
+    # C2 not provided
+    backbone_layers = {
+        'C3' : model.outputs[0],
+        'C4' : model.outputs[1],
+        'C5' : model.outputs[2]
+        }
```
Comment (hgaiser) [inline keras_retinanet/models/effnet.py:125]: Same here.

File: keras_retinanet/models/mobilenet.py [hunk 21]
```diff
@@ -106,4 +106,11 @@ def mobilenet_retinanet(num_classes, backbone='mobilenet224_1.0', inputs=None, m
     if modifier:
         backbone = modifier(backbone)
 
-    return retinanet.retinanet(inputs=inputs, num_classes=num_classes, backbone_layers=backbone.outputs, **kwargs)
+    # C2 not provided
+    backbone_layers = {
+        'C3' : backbone.outputs[0],
+        'C4' : backbone.outputs[1],
+        'C5' : backbone.outputs[2]
+        }
```
Comment (hgaiser) [inline keras_retinanet/models/mobilenet.py:114]: And here.

File: keras_retinanet/models/resnet.py [hunk 22]
```diff
@@ -109,7 +109,15 @@ def resnet_retinanet(num_classes, backbone='resnet50', inputs=None, modifier=Non
...
 
     # create the full model
-    return retinanet.retinanet(inputs=inputs, num_classes=num_classes, backbone_layers=resnet.outputs[1:], **kwargs)
+    # resnet.outputs contains 4 layers
+    backbone_layers = {
+        'C2' : resnet.outputs[0],
+        'C3' : resnet.outputs[1],
+        'C4' : resnet.outputs[2],
+        'C5' : resnet.outputs[3]
+        }
```
Comment (hgaiser) [inline keras_retinanet/models/resnet.py:118]: And here.

File: keras_retinanet/models/retinanet.py [hunk 23]
```diff
@@ -124,42 +124,65 @@ def default_regression_model(num_values, num_anchors, pyramid_feature_size=256,
     return keras.models.Model(inputs=inputs, outputs=outputs, name=name)
 
 
-def __create_pyramid_features(C3, C4, C5, feature_size=256):
+def __create_pyramid_features(C3, C4, C5, pyramid_levels, C2=None, feature_size=256):
```
Comment (hgaiser) [inline keras_retinanet/models/retinanet.py:127]: Shouldn't this just take the dict now?

File: keras_retinanet/models/retinanet.py [hunk 24]
```diff
@@ -273,13 +298,30 @@ def retinanet(
...
+
+    C3 = backbone_layers['C3']
+    C4 = backbone_layers['C4']
+    C5 = backbone_layers['C5']
+
+    if 2 in pyramid_levels:
+        if 'C2' in backbone_layers:
+            C2 = backbone_layers['C2']
+        else:
+            raise ValueError('C2 not provided by backbone model. Cannot create P2 layers.')
```
Comment (hgaiser) [inline keras_retinanet/models/retinanet.py:314]: We make the habit of putting "user readable strings" in double quotes, like this one.

File: keras_retinanet/models/retinanet.py [hunk 25]
```diff
@@ -273,13 +298,30 @@ def retinanet(
...
+
+    if 3 not in pyramid_levels or 4 not in pyramid_levels or 5 not in pyramid_levels:
+        raise ValueError('pyramid levels 3, 4, and 5 required for functionality')
```
Comment (hgaiser) [inline keras_retinanet/models/retinanet.py:317]: And this one.

Comment (mooratov) [reply keras_retinanet/models/retinanet.py]: fixed.

File: keras_retinanet/models/retinanet.py (hunk 24 above)
Comment (mooratov) [reply keras_retinanet/models/retinanet.py]: fixed.

File: keras_retinanet/models/retinanet.py (hunk 23 above)
Comment (mooratov) [reply keras_retinanet/models/retinanet.py]: Good point. Everything now works with dicts end to end.

File: keras_retinanet/models/resnet.py (hunk 22 above)
Comment (mooratov) [reply keras_retinanet/models/resnet.py]: fixed

File: keras_retinanet/models/mobilenet.py (hunk 21 above)
Comment (mooratov) [reply keras_retinanet/models/mobilenet.py]: fixed

File: keras_retinanet/models/effnet.py (hunk 20 above)
Comment (mooratov) [reply keras_retinanet/models/effnet.py]: fixed

File: keras_retinanet/models/densenet.py (hunk 19 above)
Comment (mooratov) [reply keras_retinanet/models/densenet.py]: fixed

File: keras_retinanet/bin/evaluate.py (hunk 18 above)
Comment (mooratov) [reply keras_retinanet/bin/evaluate.py]: fixed.. I am now using a python linter to hopefully prevent these things =)

File: keras_retinanet/bin/convert_model.py (hunk 17 above)
Comment (mooratov) [reply keras_retinanet/bin/convert_model.py]: yep. Removed now, thanks.

**Comment by mooratov** [comment]
Thanks again for the feedback. I've made the requested changes, and have finally started using a linter on my editor, which helped me catch a few more stylistic lapses as well as a few genuine bugs. I've confirmed that train.py, evaluate.py, convert_model.py, and debug.py all work as intended now. 

There are a few issues that can potentially arise however if users are not careful with configuring anchor parameters and pyramid levels. I am not sure to what extent you want me to foolproof this type of stuff, but I thought I would bring them to your attention at the very least: 

- `pyramid_levels` and `anchor_params.sizes` lists must have the same length or the code won't work. This can probably be fixed with an assert statement in both the `retinanet` and `retinanet_bbox` functions in `retinanet.py` to ensure they have the same length.

- On a similar note, `anchor_params.strides` and `anchor_params.sizes` should have the same length ( this is an issue in the code independent of my PR ). Can be fixed with an assert in the config.py function that reads these anchor parameters (or elsewhere) 

- If you use `evaluate.py` on a model that was trained with non-standard pyramid levels, you MUST provide the `config.ini` file to set the pyramid levels properly, or else it will crash ungracefully due to tensor dimensionality mismatch. The crash occurs when `eval.py` runs `model.predict_on_batch(..)`. This one would be more tricky for me to try fixing right now.

**Comment by hgaiser** [comment]
Good points, I think the asserts are relatively easy to add and can provide somewhat of a security against wrong configurations. The third point is indeed a bit more tricky, I would leave that as it is for now.

**Comment by mooratov** [comment]
ok all done!

File: keras_retinanet/utils/config.py [hunk 26]
```diff
@@ -35,6 +35,9 @@ def read_config_file(config_path):
     assert config_keys <= default_keys, \
         "Malformed config file. These keys are not valid: {}".format(config_keys - default_keys)
 
+    if 'pyramid_levels' in config:
+        assert('P' in config['pyramid_levels']), "pyramid levels specified by P key"
```
Comment (hgaiser) [inline keras_retinanet/utils/config.py:39]: Does this work? Afaik there is no `'P'` key in the pyramid levels config right? Did you intend to search for a key that starts with `'P'`?

Comment (mooratov) [reply keras_retinanet/utils/config.py:39]: Well previously there was no options at all for pyramid levels config. The way I implemented formatting the config files is as follows:

```
[anchor_parameters]
...
[pyramid_levels]
P = 2 3 4 5 6 7
```

You are right that this may be opaque (would need to modify the example config.ini template found elsewhere on the repo) , so open to suggestion here if you have a preferred way.

Comment (hgaiser) [reply keras_retinanet/utils/config.py:39]: How about something a bit more descriptive, like `'levels'`?

Comment (mooratov) [reply keras_retinanet/utils/config.py]: alright, done

Comment (mooratov) [reply keras_retinanet/utils/config.py]: hello, just wanted to ping to see if there's any other requested changes.

**Comment by hgaiser** [comment]
Looks good to me, thanks again for your contribution :)

**Comment by MishallSwain** [comment]
Unable to train after this commit. Please help.

Traceback (most recent call last):
  File "train.py", line 540, in <module>
    main()
  File "train.py", line 500, in main
    config=args.config
  File "train.py", line 114, in create_models
    model          = model_with_weights(backbone_retinanet(num_classes, num_anchors=num_anchors, modifier=modifier), weights=weights, skip_mismatch=True)
  File "../../keras_retinanet/models/resnet.py", line 38, in retinanet
    return resnet_retinanet(*args, backbone=self.backbone, **kwargs)
  File "../../keras_retinanet/models/resnet.py", line 99, in resnet_retinanet
    resnet = keras_resnet.models.ResNet50(inputs, include_top=False, freeze_bn=True)
  File "/home/conawarehouse/.local/lib/python3.7/site-packages/keras_resnet/models/_2d.py", line 188, in ResNet50
    return ResNet(inputs, blocks, numerical_names=numerical_names, block=keras_resnet.blocks.bottleneck_2d, include_top=include_top, classes=classes, *args, **kwargs)
  File "/home/conawarehouse/.local/lib/python3.7/site-packages/keras_resnet/models/_2d.py", line 66, in ResNet
    x = keras_resnet.layers.BatchNormalization(axis=axis, epsilon=1e-5, freeze=freeze_bn, name="bn_conv1")(x)
  File "/anaconda/envs/tensorflow_p36/lib/python3.7/site-packages/tensorflow/python/keras/engine/base_layer.py", line 922, in __call__
    outputs = call_fn(cast_inputs, *args, **kwargs)
  File "/anaconda/envs/tensorflow_p36/lib/python3.7/site-packages/tensorflow/python/autograph/impl/api.py", line 265, in wrapper
    raise e.ag_error_metadata.to_exception(e)
TypeError: in user code:

    /home/conawarehouse/.local/lib/python3.7/site-packages/keras_resnet/layers/_batch_normalization.py:17 call  *
        return super(BatchNormalization, self).call(training=(not self.freeze), *args, **kwargs)

    TypeError: type object got multiple values for keyword argument 'training'

**Comment by mooratov** [comment]
Hi, I believe that this is related to the Keras 2.4.0 release that happened
a few days ago. It appears to break some fundamental compatibility
(possibly because it forces installation of keras-resnet==0.2.0

pip uninstall keras
pip install keras==2.3.0

fixed it for me.
Not a satisfying fix, but also this is unrelated to the newly merged PR
